src/compression/snappy.rs

```diff
@@ -1,18 +1,37 @@
 use std::io::{self, Read};
 
 use byteorder::{BigEndian, ByteOrder};
-use snappy;
+use snap;
 
 use error::{Result, Error};
 
 pub fn compress(src: &[u8]) -> Result<Vec<u8>> {
-    Ok(snappy::compress(src))
+    let mut buf = vec![0; snap::max_compress_len(src.len())];
+
+    snap::Encoder::new()
+        .compress(src, &mut buf)
+        .map(|len| {
+                 buf.truncate(len);
+
+                 buf
+             })
+        .map_err(|err| Error::InvalidSnappy(err))
 }
 
 fn uncompress_to(src: &[u8], dst: &mut Vec<u8>) -> Result<()> {
-    snappy::uncompress_to(src, dst)
-        .map(|_| ())
-        .map_err(|_| Error::InvalidInputSnappy)
+    snap::decompress_len(src)
+        .and_then(|min_len| {
+            let off = dst.len();
```

Comment:
Parsing the header is very fast. It's a single variable integer read and a couple branches.

In any case, the only way around this while using a single call to `decompress` would be:

1. `snap` provided a `decompress` method that accepted a `&mut Vec<u8>` and did the resizing for you.
2. `snap` provided a `decompress` method that accepted the parsed header and "trusted" that information.

(2) doesn't seem wise and is a potential footgun. (1) is potentially interesting, but I would block that addition unless there was a benchmark showing compelling evidence that it is worth it.

Another approach would be to implement a streaming raw decoder in `snap`, but that's basically a rewrite, and most use cases are satified by just using the snappy frame format (which I assume doesn't work here).